lib/operations/execute_operation.js

```diff
@@ -45,18 +59,14 @@ function executeOperation(topology, operation, callback) {
       session = topology.startSession({ owner });
       operation.session = session;
     } else if (operation.session.hasEnded) {
-      throw new MongoError('Use of expired sessions is not permitted');
+      callback(new MongoError('Use of expired sessions is not permitted'));
+      return result;
     }
-  }
-
-  let result;
-  if (typeof callback !== 'function') {
-    result = new Promise((resolve, reject) => {
-      callback = (err, res) => {
-        if (err) return reject(err);
-        resolve(res);
-      };
-    });
+  } else if (operation.session && !topology.hasSessionSupport()) {
```

Comment:
another nit: I would pull this out into a very explicit block right below the `selectServerForSessionSupport` call, just to make it very clear that we are looking for this particular case:

```
  if (isUnifiedTopology(topology) && topology.shouldCheckForSessionSupport()) {
    selectServerForSessionSupport(topology, operation, callback);
    return result;
  }
 
  if (operation.session && !topology.hasSessionSupport()) {
    callback(new MongoError('Current topology does not support sessions'));
    return result;
  }

  .. else everything goes on as expected
```

Reply:
I cleaned up the `else if` condition to make it a bit easier to read the control flow, I think that might make it more quickly understandable what is being checked here.